libraries/AP_WindVane/AP_WindVane.cpp

```diff
@@ -0,0 +1,334 @@
+
+    // @Param: ANA_PIN
+    // @DisplayName: Analog input
+    // @Description: Analog input pin to read as Wind vane sensor pot
+    // @Values: 11:Pixracer,13:Pixhawk ADC4,14:Pixhawk ADC3,15:Pixhawk ADC6,15:Pixhawk2 ADC,50:PixhawkAUX1,51:PixhawkAUX2,52:PixhawkAUX3,53:PixhawkAUX4,54:PixhawkAUX5,55:PixhawkAUX6,103:Pixhawk SBUS
+    // @User: Standard
+    AP_GROUPINFO("ANA_PIN", 3, AP_WindVane, _analog_pin_no, WINDVANE_DEFAULT_PIN),
+
+    // @Param: ANA_V_MIN
+    // @DisplayName: Analog minumum voltage
+    // @Description: Minimum analog voltage read by wind vane
+    // @Units: V
+    // @Increment: 0.01
+    // @Range: 0 5.0
+    // @User: Standard
+    AP_GROUPINFO("ANA_V_MIN", 4, AP_WindVane, _analog_volt_min, 0.0f),
+
+    // @Param: ANA_V_MAX
+    // @DisplayName: Analog maximum voltage
+    // @Description: Minimum analog voltage read by wind vane
+    // @Units: V
+    // @Increment: 0.01
+    // @Range: 0 5.0
+    // @User: Standard
+    AP_GROUPINFO("ANA_V_MAX", 5, AP_WindVane, _analog_volt_max, 3.3f),
+
+    // @Param: ANA_OF_HD
+    // @DisplayName: Analog headwind offset
+    // @Description: Angle offset when windvane is indicating a headwind, ie 0 degress relative to vehicle
+    // @Units: deg
+    // @Increment: 1
+    // @Range: 0 360
+    // @User: Standard
+    AP_GROUPINFO("ANA_OF_HD", 6, AP_WindVane, _analog_head_bearing_offset, 0.0f),
+
+    // @Param: VANE_FILT
+    // @DisplayName: Wind vane low pass filter frequency
+    // @Description: Wind vane low pass filter frequency, a value of -1 disables filter
+    // @Units: Hz
+    // @User: Standard
+    AP_GROUPINFO("VANE_FILT", 7, AP_WindVane, _vane_filt_hz, 0.5f),
+
+    // @Param: CAL
+    // @DisplayName: Wind vane calibration start
+    // @Description: Start wind vane calibration by setting this to 1
+    // @Values: 0:None, 1:Calibrate
+    // @User: Standard
+    AP_GROUPINFO("CAL", 8, AP_WindVane, _calibration, 0),
+
+    // @Param: ANA_DZ
+    // @DisplayName: Analog potentiometer dead zone
+    // @Description: Analog potentiometer mechanical dead zone
+    // @Units: deg
+    // @Increment: 1
+    // @Range: 0 360
+    // @User: Standard
+    AP_GROUPINFO("ANA_DZ", 9, AP_WindVane, _analog_deadzone, 0),
+
+    AP_GROUPEND
+};
+
+// constructor
+AP_WindVane::AP_WindVane()
+{
+    AP_Param::setup_object_defaults(this, var_info);
+#if CONFIG_HAL_BOARD == HAL_BOARD_SITL
+    if (_singleton) {
+        AP_HAL::panic("Too many Wind Vane sensors");
+    }
+#endif
+    _singleton = this;
+}
+
+AP_WindVane::~AP_WindVane()
+{
+}
+
+/*
+ * Get the AP_WindVane singleton
+ */
+AP_WindVane *AP_WindVane::get_singleton()
+{
+    return _singleton;
+}
+
+// return true if wind vane is enabled
+bool AP_WindVane::enabled() const
+{
+    return (_type != WINDVANE_NONE);
+}
+
+// Initialize the Wind Vane object and prepare it for use
+void AP_WindVane::init()
+{
+    // a pin for reading the Wind Vane voltage
+    windvane_analog_source = hal.analogin->channel(_analog_pin_no);
```

Comment:
shouldn't this done only when type is analog ?

Reply:
yes, and it is also called on every iteration.  I will remove the Init function completely.  We will need to add it back when we add wind speed support but that's no problem.

Reply:
This was wrong. This is what actually set the pointer. You need to call channel() once to get an object to work on. Setting the default pin is worth while, but you really do need to try and call channel(). With the change you pushed it will now do nullpointer dereferences.

Reply:
@WickedShell, yes, just discovered that little mistake in testing.  I've put the init back and added a RebootRequired to the param description.

Reply:
If you keep the set_pin() then a reboot isn't required to change the pin.

Reply:
damn, can't get this right :-).